src/parsers/manifest/dash/parse_representation_index.ts

```diff
@@ -0,0 +1,161 @@
+ * @param {Array.<Object>} representation
+ * @param {Object} representationInfos
+ * @returns {Array.<Object>}
+ */
+export default function parseRepresentationIndex(
+  representation : IRepresentationIntermediateRepresentation,
+  representationInfos : IRepresentationInfos
+) : IRepresentationIndex {
+  const representationBaseURLs = resolveBaseURLs(representationInfos.baseURLs,
+                                                 representation.children.baseURLs);
+  const { aggressiveMode,
+          availabilityTimeOffset,
+          manifestBoundsCalculator,
+          isDynamic,
+          end: periodEnd,
+          start: periodStart,
+          receivedTime,
+          timeShiftBufferDepth,
+          unsafelyBaseOnPreviousRepresentation } = representationInfos;
+  const context = { aggressiveMode,
+                    availabilityTimeOffset,
+                    unsafelyBaseOnPreviousRepresentation,
+                    manifestBoundsCalculator,
+                    isDynamic,
+                    periodEnd,
+                    periodStart,
+                    receivedTime,
+                    representationBaseURLs,
+                    representationBitrate: representation.attributes.bitrate,
+                    representationId: representation.attributes.id,
+                    timeShiftBufferDepth };
+  let representationIndex : IRepresentationIndex;
+  if (representation.children.segmentBase !== undefined) {
+    const { segmentBase } = representation.children;
+    context.availabilityTimeOffset =
+      representationInfos.availabilityTimeOffset +
+      extractMinimumAvailabilityTimeOffset(representation.children.baseURLs) +
+      (segmentBase.availabilityTimeOffset ?? 0);
+    representationIndex = new BaseRepresentationIndex(segmentBase, context);
+  } else if (representation.children.segmentList !== undefined) {
+    const { segmentList } = representation.children;
+    representationIndex = new ListRepresentationIndex(segmentList, context);
+  } else if (representation.children.segmentTemplate !== undefined ||
+             representationInfos.parentSegmentTemplates.length > 0)
+  {
+    const segmentTemplates = representationInfos.parentSegmentTemplates.slice();
+    const childSegmentTemplate = representation.children.segmentTemplate;
+    if (childSegmentTemplate !== undefined) {
+      segmentTemplates.push(childSegmentTemplate);
+    }
+    const segmentTemplate =
+      objectAssign({},
+                   ...segmentTemplates as [IParsedSegmentTemplate] /* Ugly TS Hack */);
+    context.availabilityTimeOffset =
+      representationInfos.availabilityTimeOffset +
+      extractMinimumAvailabilityTimeOffset(representation.children.baseURLs) +
+      (segmentTemplate.availabilityTimeOffset ?? 0);
+    representationIndex = segmentTemplate.timelineParser !== undefined ?
+      new TimelineRepresentationIndex(segmentTemplate,
+                                      segmentTemplate.timelineParser,
```

Comment:
Maybe it is not clear for everybody here that Typescript understands that the timelineParser is defined, and thus the TimelineRepresentationIndex takes the segmentTemplate, and one of the attribute of it, instead of taking just the segmentTemplate with the doubt on the timelineParser presence inside of it.

Wouldn't it be better to deconstruct the timelineParser, then check if it is defined, and use it as an argument. Maybe it is clearer.

Reply:
Done